File: isegoria/src/main/java/com/eulersbridge/isegoria/feed/events/EventDetailActivity.java
package com.eulersbridge.isegoria.feed.events;

import android.content.Intent;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.RectShape;
import android.os.Bundle;
import android.provider.CalendarContract;
import android.support.v7.app.AppCompatActivity;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;

import com.bumptech.glide.Priority;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.eulersbridge.isegoria.GlideApp;
import com.eulersbridge.isegoria.IsegoriaApp;
import com.eulersbridge.isegoria.network.api.models.Event;
import com.eulersbridge.isegoria.network.api.models.Position;
import com.eulersbridge.isegoria.network.api.models.Ticket;
import com.eulersbridge.isegoria.network.api.models.User;
import com.eulersbridge.isegoria.profile.ProfileOverviewFragment;
import com.eulersbridge.isegoria.util.Constants;
import com.eulersbridge.isegoria.util.Strings;
import com.eulersbridge.isegoria.util.network.SimpleCallback;
import com.eulersbridge.isegoria.util.transformation.TintTransformation;
import com.eulersbridge.isegoria.R;

import org.parceler.Parcels;

import retrofit2.Response;

public class EventDetailActivity extends AppCompatActivity {

    private float dpWidth;
    private IsegoriaApp isegoriaApp;

    private TableLayout eventContactTableLayout;

    private Event event;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.event_detail_activity);

        getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);

        findViewById(R.id.back_button).setOnClickListener(view -> onBackPressed());

        eventContactTableLayout = findViewById(R.id.event_details_table_layout);

        Button addToCalendar = findViewById(R.id.event_button_add_to_calendar);
        addToCalendar.setOnClickListener(view -> addToCalendar());

        DisplayMetrics displayMetrics = getResources().getDisplayMetrics();
        dpWidth = displayMetrics.widthPixels / displayMetrics.density;

        event = Parcels.unwrap(getIntent().getParcelableExtra(Constants.ACTIVITY_EXTRA_EVENT));

        isegoriaApp = (IsegoriaApp)getApplication();

        if (event != null) {
            populateContent(event);

            ImageView eventImageView = findViewById(R.id.event_image);

            GlideApp.with(this)
                    .load(event.getPhotoUrl())
                    .priority(Priority.HIGH)
                    .transform(new TintTransformation())
                    .placeholder(R.color.lightGrey)
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(eventImageView);
        }
    }

    private void populateContent(final Event event) {
        runOnUiThread(() -> {
            TextView eventTitleField = findViewById(R.id.event_title);
            eventTitleField.setText(event.name);

            TextView eventTime = findViewById(R.id.event_time);
            eventTime.setText(Strings.fromTimestamp(this, event.date));

            TextView eventLocationLine1 = findViewById(R.id.event_location);
            eventLocationLine1.setText(event.location);

            TextView eventsTextField = findViewById(R.id.event_details);
            eventsTextField.setText(event.description);

            addCandidate(event.organizerEmail);
        });
    }

    private void addToCalendar() {
        Intent intent = new Intent(Intent.ACTION_INSERT)
                .setData(CalendarContract.Events.CONTENT_URI)
                .putExtra(CalendarContract.EXTRA_EVENT_BEGIN_TIME, event.date)

                // Make event 1 hour long (add an hour in in milliseconds to event start)
                .putExtra(CalendarContract.EXTRA_EVENT_END_TIME, event.date + 60 * 60 * 1000)
                .putExtra(CalendarContract.Events.ALL_DAY, false)

                .putExtra(CalendarContract.Events.TITLE, event.name)
                .putExtra(CalendarContract.Events.DESCRIPTION, event.description)
                .putExtra(CalendarContract.Events.EVENT_LOCATION, event.location)
                .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

        if (intent.resolveActivity(getPackageManager()) != null) {
            isegoriaApp.getMainActivity().startActivity(intent);
        }
    }

    private void addCandidate(String email) {
        TableRow tr;

        tr = new TableRow(this);
        TableRow.LayoutParams rowParams = new TableRow.LayoutParams(TableRow.LayoutParams.MATCH_PARENT, TableRow.LayoutParams.WRAP_CONTENT);
        tr.setLayoutParams(rowParams);
        tr.setPadding(10, 10, 0, 10);

        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.HORIZONTAL);

        TextView contactTextView = new TextView(this);
        contactTextView.setTypeface(null, Typeface.BOLD);
        contactTextView.setText(getString(R.string.event_detail_activity_organiser, email));
        layout.addView(contactTextView);
        tr.addView(layout);

        eventContactTableLayout.addView(tr);
    }

    public void addCandidate(int userId, int ticketId, int positionId, int candidateId,
                             String firstName, String lastName) {
        addTableRow(ticketId, null, userId, "GRN", "#4FBE3E", String.format("%s %s", firstName, lastName), "", positionId);
    }

    private void addTableRow(int ticketId, User user, final int userId, String partyAbr,
                             String colour, String candidateName,
                             String candidatePosition, int positionId) {
        TableRow tr;

        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.HORIZONTAL);

        tr = new TableRow(this);
        TableRow.LayoutParams rowParams = new TableRow.LayoutParams(TableRow.LayoutParams.MATCH_PARENT, TableRow.LayoutParams.WRAP_CONTENT);
        tr.setLayoutParams(rowParams);
        tr.setPadding(0, 10, 0, 10);

        ImageView candidateProfileView = new ImageView(this);
        candidateProfileView.setLayoutParams(new TableRow.LayoutParams(TableRow.LayoutParams.WRAP_CONTENT, TableRow.LayoutParams.WRAP_CONTENT));
        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(80, 80);
        candidateProfileView.setLayoutParams(layoutParams);
        candidateProfileView.setScaleType(ImageView.ScaleType.CENTER_CROP);
        candidateProfileView.setPadding(10, 0, 10, 0);

        GlideApp.with(this)
                .load(user.profilePhotoURL)
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(candidateProfileView);

        ImageView candidateProfileImage = new ImageView(this);
        candidateProfileImage.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT, Gravity.END));
        candidateProfileImage.setScaleType(ImageView.ScaleType.CENTER_CROP);
        candidateProfileImage.setImageResource(R.drawable.profilelight);
        candidateProfileImage.setPadding(10, 0, 10, 0);
        candidateProfileImage.setOnClickListener(view -> {

            ProfileOverviewFragment profileOverviewFragment = new ProfileOverviewFragment();
            Bundle args = new Bundle();
            args.putInt(Constants.FRAGMENT_EXTRA_PROFILE_ID, userId);

            getSupportFragmentManager().beginTransaction()
                    .setReorderingAllowed(true)
                    .addToBackStack(null)
                    .replace(R.id.container, profileOverviewFragment)
                    .commit();
        });

        TextView textViewParty = new TextView(this);
        textViewParty.setTextColor(Color.parseColor("#FFFFFF"));
        textViewParty.setTextSize(TypedValue.COMPLEX_UNIT_DIP, 12.0f);
        textViewParty.setText(partyAbr);
        textViewParty.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT));
        textViewParty.setGravity(Gravity.CENTER);
        textViewParty.setTypeface(null, Typeface.BOLD);

        isegoriaApp.getAPI().getTicket(ticketId).enqueue(new SimpleCallback<Ticket>() {
            @Override
            public void handleResponse(Response<Ticket> response) {
                Ticket ticket = response.body();
                if (ticket != null) {
                    textViewParty.setText(ticket.code);
                    textViewParty.setBackgroundColor(Color.parseColor(ticket.getColour()));
                }
            }
        });

        RectShape rect = new RectShape();
        ShapeDrawable rectShapeDrawable = new ShapeDrawable(rect);
        Paint paint = rectShapeDrawable.getPaint();
        paint.setColor(Color.parseColor(colour));
        paint.setStyle(Paint.Style.FILL_AND_STROKE);
        paint.setStrokeWidth(5);

        LinearLayout partyLayout = new LinearLayout(this);
        partyLayout.setOrientation(LinearLayout.VERTICAL);
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                80, 40);
        params.gravity = Gravity.CENTER_VERTICAL;
        partyLayout.setLayoutParams(params);
        //partyLayout.setBackgroundDrawable(rectShapeDrawable);
        partyLayout.addView(textViewParty);

        TextView textViewCandidate = new TextView(this);
        textViewCandidate.setTextColor(Color.parseColor("#3A3F43"));
        textViewCandidate.setTextSize(TypedValue.COMPLEX_UNIT_DIP, 16.0f);
        textViewCandidate.setText(candidateName);
        textViewCandidate.setPadding(10, 0, 10, 0);
        textViewCandidate.setGravity(Gravity.START);

        TextView textViewPosition = new TextView(this);
        textViewPosition.setTextColor(Color.parseColor("#3A3F43"));
        textViewPosition.setTextSize(TypedValue.COMPLEX_UNIT_DIP, 12.0f);
        textViewPosition.setText(candidatePosition);
        textViewPosition.setPadding(10, 0, 10, 0);
        textViewPosition.setGravity(Gravity.START);

        isegoriaApp.getAPI().getPosition(positionId).enqueue(new SimpleCallback<Position>() {
            @Override
            public void handleResponse(Response<Position> response) {
                Position position = response.body();
                if (position != null) textViewPosition.setText(position.name);
            }
        });

        View dividerView = new View(this);
        dividerView.setLayoutParams(new TableRow.LayoutParams(TableRow.LayoutParams.MATCH_PARENT, 1));
        dividerView.setBackgroundColor(Color.parseColor("#676475"));

        RelativeLayout relLayoutMaster = new RelativeLayout(this);
        TableRow.LayoutParams relLayoutMasterParam = new TableRow.LayoutParams((int)dpWidth, TableRow.LayoutParams.WRAP_CONTENT);
        relLayoutMaster.setLayoutParams(relLayoutMasterParam);

        RelativeLayout.LayoutParams relativeParamsLeft = new RelativeLayout.LayoutParams(TableRow.LayoutParams.MATCH_PARENT, TableRow.LayoutParams.WRAP_CONTENT);
        relativeParamsLeft.addRule(RelativeLayout.ALIGN_PARENT_LEFT);

        RelativeLayout.LayoutParams relativeParamsRight = new RelativeLayout.LayoutParams(TableRow.LayoutParams.MATCH_PARENT, TableRow.LayoutParams.WRAP_CONTENT);
        relativeParamsRight.addRule(RelativeLayout.ALIGN_PARENT_RIGHT);

        LinearLayout linLayout = new LinearLayout(this);
        linLayout.setOrientation(LinearLayout.VERTICAL);
        linLayout.addView(textViewCandidate);
        linLayout.addView(textViewPosition);

        LinearLayout linLayout2 = new LinearLayout(this);
        linLayout2.setOrientation(LinearLayout.VERTICAL);
        linLayout2.addView(candidateProfileImage);
        linLayout2.setGravity(Gravity.END);
        linLayout2.setLayoutParams(relativeParamsRight);

        layout.addView(candidateProfileView);
        layout.addView(partyLayout);
        layout.addView(linLayout);
        layout.setLayoutParams(relativeParamsLeft);

        relLayoutMaster.addView(layout);
        relLayoutMaster.addView(linLayout2);

        tr.addView(relLayoutMaster);

        eventContactTableLayout.addView(tr);
        eventContactTableLayout.addView(dividerView);
    }
}

File: isegoria/src/main/java/com/eulersbridge/isegoria/feed/events/detail/EventDetailActivity.kt
package com.eulersbridge.isegoria.feed.events.detail

import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.RectShape
import android.os.Bundle
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.widget.*
import androidx.core.os.bundleOf
import com.bumptech.glide.Priority
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.eulersbridge.isegoria.*
import com.eulersbridge.isegoria.feed.events.ACTIVITY_EXTRA_EVENT
import com.eulersbridge.isegoria.network.api.model.Event
import com.eulersbridge.isegoria.network.api.model.User
import com.eulersbridge.isegoria.profile.ProfileOverviewFragment
import com.eulersbridge.isegoria.util.extension.observe
import com.eulersbridge.isegoria.util.extension.toDateString
import com.eulersbridge.isegoria.util.transformation.TintTransformation
import dagger.android.support.DaggerAppCompatActivity
import kotlinx.android.synthetic.main.event_detail_activity.*
import javax.inject.Inject

class EventDetailActivity : DaggerAppCompatActivity() {

    private val dpWidth: Float by lazy {
        val displayMetrics = resources.displayMetrics
        displayMetrics.widthPixels / displayMetrics.density
    }

    @Inject
    lateinit var modelFactory: ViewModelProvider.Factory
    lateinit var viewModel: EventDetailViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.event_detail_activity)

        window.decorView.systemUiVisibility =
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN

        viewModel = ViewModelProviders.of(this, modelFactory)[EventDetailViewModel::class.java]

        backButton.setOnClickListener { onBackPressed() }
        addToCalendarButton.setOnClickListener { viewModel.addToCalendar(this) }

        val event = intent.getParcelableExtra<Event>(ACTIVITY_EXTRA_EVENT)
        viewModel.event.value = event

        if (event != null) {
            populateContent(event)

            GlideApp.with(this)
                .load(event.getPhotoUrl())
                .priority(Priority.HIGH)
                .transform(TintTransformation())
                .placeholder(R.color.lightGrey)
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(eventImageView)
        }
    }

    private fun populateContent(event: Event) {
        runOnUiThread {
            titleTextView.text = event.name
            timeTextView.text = event.createdDate.toDateString(this)
            locationTextView.text = event.location
            detailsTextView.text = event.description

            event.organizerEmail?.let {
                addCandidate(it)
            }
        }
    }

    private fun addCandidate(email: String) {
        val tr = TableRow(this)
        tr.apply {
            layoutParams = TableRow.LayoutParams(
                TableRow.LayoutParams.MATCH_PARENT,
                TableRow.LayoutParams.WRAP_CONTENT
            )

            setPadding(10, 10, 0, 10)
        }

        val layout = LinearLayout(this)
        layout.orientation = LinearLayout.HORIZONTAL

        val contactTextView = TextView(this)
        contactTextView.apply {
            setTypeface(null, Typeface.BOLD)
            text = getString(R.string.event_detail_activity_organiser, email)
        }

        layout.addView(contactTextView)
        tr.addView(layout)

        tableLayout.addView(tr)
    }

    fun addCandidate(
        userId: Int, ticketId: Int, positionId: Int, candidateId: Int,
        firstName: String, lastName: String
    ) {
        addTableRow(
            ticketId,
            null,
            userId,
            "GRN",
            "#4FBE3E",
            "$firstName $lastName",
            "",
            positionId
        )
    }

    private fun addTableRow(
        ticketId: Int, user: User?, userId: Int, partyAbr: String,
        colour: String, candidateName: String,
        candidatePosition: String, positionId: Int
    ) {
        val tr = TableRow(this)

        val layout = LinearLayout(this)
        layout.orientation = LinearLayout.HORIZONTAL

        val rowParams = TableRow.LayoutParams(
            TableRow.LayoutParams.MATCH_PARENT,
            TableRow.LayoutParams.WRAP_CONTENT
        )
        tr.layoutParams = rowParams
        tr.setPadding(0, 10, 0, 10)

        val candidateProfileView = ImageView(this)
        candidateProfileView.apply {
            layoutParams = TableRow.LayoutParams(
                TableRow.LayoutParams.WRAP_CONTENT,
                TableRow.LayoutParams.WRAP_CONTENT
            )

            scaleType = ImageView.ScaleType.CENTER_CROP
            setPadding(10, 0, 10, 0)
            layoutParams = LinearLayout.LayoutParams(80, 80)
        }

        GlideApp.with(this)
            .load(user?.profilePhotoURL)
            .transition(DrawableTransitionOptions.withCrossFade())
            .into(candidateProfileView)

        val candidateProfileImage = ImageView(this)
        candidateProfileImage.apply {

            layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT,
                Gravity.END.toFloat()
            )

            scaleType = ImageView.ScaleType.CENTER_CROP

            setImageResource(R.drawable.profile_light)
            setPadding(10, 0, 10, 0)
            setOnClickListener {
                val profileFragment = ProfileOverviewFragment()
                profileFragment.arguments = bundleOf(FRAGMENT_EXTRA_PROFILE_ID to userId)

                supportFragmentManager.beginTransaction()
                    .setReorderingAllowed(true)
                    .addToBackStack(null)
                    .replace(R.id.container, profileFragment)
                    .commit()
            }
        }

        val textViewParty = TextView(this)
        textViewParty.apply {
            setTextColor(Color.parseColor("#FFFFFF"))
            setTextSize(TypedValue.COMPLEX_UNIT_DIP, 12.0f)
            setTypeface(null, Typeface.BOLD)

            text = partyAbr
            gravity = Gravity.CENTER

            layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT
            )
        }

        observe(viewModel.getTicket(ticketId.toLong())) {
            if (it != null)
                textViewParty.apply {
                    text = it.code
                    setBackgroundColor(Color.parseColor(it.getColour()))
                }
        }

        val rect = RectShape()
        val rectShapeDrawable = ShapeDrawable(rect)
        val paint = rectShapeDrawable.paint
        paint.color = Color.parseColor(colour)
        paint.style = Paint.Style.FILL_AND_STROKE
        paint.strokeWidth = 5f

        val partyLayout = LinearLayout(this)
        partyLayout.orientation = LinearLayout.VERTICAL
        val params = LinearLayout.LayoutParams(
            80, 40
        )
        params.gravity = Gravity.CENTER_VERTICAL
        partyLayout.layoutParams = params
        //partyLayout.setBackgroundDrawable(rectShapeDrawable);
        partyLayout.addView(textViewParty)

        val textViewCandidate = TextView(this)
        textViewCandidate.apply {
            setTextColor(Color.parseColor("#3A3F43"))
            setTextSize(TypedValue.COMPLEX_UNIT_DIP, 16.0f)
            text = candidateName
            setPadding(10, 0, 10, 0)
            gravity = Gravity.START
        }

        val textViewPosition = TextView(this)
        textViewPosition.apply {
            setTextColor(Color.parseColor("#3A3F43"))
            setTextSize(TypedValue.COMPLEX_UNIT_DIP, 12.0f)
            text = candidatePosition
            setPadding(10, 0, 10, 0)
            gravity = Gravity.START
        }

        observe(viewModel.getPosition(positionId.toLong())) {
            textViewPosition.text = it?.name
        }

        val dividerView = View(this)
        dividerView.layoutParams = TableRow.LayoutParams(TableRow.LayoutParams.MATCH_PARENT, 1)
        dividerView.setBackgroundColor(Color.parseColor("#676475"))

        val relLayoutMaster = RelativeLayout(this)
        val relLayoutMasterParam =
            TableRow.LayoutParams(dpWidth.toInt(), TableRow.LayoutParams.WRAP_CONTENT)
        relLayoutMaster.layoutParams = relLayoutMasterParam

        val relativeParamsLeft = RelativeLayout.LayoutParams(
            TableRow.LayoutParams.MATCH_PARENT,
            TableRow.LayoutParams.WRAP_CONTENT
        )
        relativeParamsLeft.addRule(RelativeLayout.ALIGN_PARENT_LEFT)

        val relativeParamsRight = RelativeLayout.LayoutParams(
            TableRow.LayoutParams.MATCH_PARENT,
            TableRow.LayoutParams.WRAP_CONTENT
        )
        relativeParamsRight.addRule(RelativeLayout.ALIGN_PARENT_RIGHT)

        val linLayout = LinearLayout(this)
        linLayout.orientation = LinearLayout.VERTICAL
        linLayout.addView(textViewCandidate)
        linLayout.addView(textViewPosition)

        val linLayout2 = LinearLayout(this)
        linLayout2.apply {
            orientation = LinearLayout.VERTICAL
            addView(candidateProfileImage)
            gravity = Gravity.END
        }
        linLayout2.layoutParams = relativeParamsRight

        layout.addView(candidateProfileView)
        layout.addView(partyLayout)
        layout.addView(linLayout)
        layout.layoutParams = relativeParamsLeft

        relLayoutMaster.addView(layout)
        relLayoutMaster.addView(linLayout2)

        tr.addView(relLayoutMaster)

        tableLayout.addView(tr)
        tableLayout.addView(dividerView)
    }
}
